src/functor.jl

```diff
@@ -169,7 +169,54 @@ function _default_walk(f, x)
   func, re = functor(x)
   re(map(f, func))
 end
-_default_walk(f, ::Nothing, ::Nothing) = nothing
+_default_walk(_, ::Nothing, ::Nothing) = nothing
+
+# Side effects only, saves a restructure
+function _foreach_walk(f, x)
+  foreach(f, children(x))
+  return x
+end
+
+### WARNING: the following is unstable internal functionality. Use at your own risk!
+# Wrapper over an IdDict which only saves values with a stable object identity
+struct Cache{K,V}
+  inner::IdDict{K,V}
+end
+Cache() = Cache(IdDict())
+
+iscachesafe(x) = !isbits(x) && ismutable(x)
+# Functionally immutable and observe value semantics, but still `ismutable` and not `isbits`
+iscachesafe(::Union{String,Symbol}) = false
+# For varargs
+iscachesafe(xs::Tuple) = all(iscachesafe, xs)
+Base.get!(f, c::Cache, x) = iscachesafe(x) ? get!(f, c.inner, x) : f()
+
+# Passthrough used to disable caching (e.g. when passing `cache=false`)
+struct NoCache end
+Base.get!(f, ::NoCache, _) = f()
```

Comment:
Style not logic, but could this be more tidily done as a `safeget!` function, rather than types on which Base's functions act? There is I think exactly one line which calls this `get!` on these types.

Reply:
I had exactly this in an earlier draft, but unfortunately it's a breaking change (i.e. tests fail) because the default before was to unconditionally trust IdDict/objectid/===. So value types were being deduped via the cache. If we decide to call that a bug, then I can put this suggestion in place.

Another theoretical benefit of not hard-coding `usecache` would be allowing users to customize caching behaviour on a case-by-case basis. I haven't thought of a scenario that would require such a degree of customization yet (e.g. `TiedArrays` could simply overload `usecache`), so we can cross that bridge when we come to it.

Reply:
Right, my suggestion still leaves `usecache` as an overloadable function. It's not intended as a change of logic, just more compact & fewer types?

Testing `isbits` etc. seems at all seems like a potentially breaking change, as it changes how many times `f` will be called (e.g. on numbers in a tuple). Maybe it depends how firmly you believe `f` must be pure... `fcollect` is not I think.

Reply:
Checking for `isbits` and `ismutable` is a hedge against an impure `f`. IMO users would be more surprised if `fmap(println, (1,2,3,1,2,3))` printed `1\n2\n3\n` as it currently does.

Reply:
Yes. I'm in favour of some such check. But perhaps this can be factored out, and we can discuss it this one has the right rules. (Should it be any or all?) This seems largely orthogonal.

Reply:
So to clarify, yea to calling the current behaviour a bug and unconditionally calling `usecache` if it helps get rid of the structs?